lib/pos_blockchain.go

```diff
@@ -76,17 +76,33 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
 		return false, false, nil, errors.Wrap(err, "processBlockPoS: Problem applying new tip: ")
 	}
 
-	// If the incoming block is not applied as the new tip, we return early.
-	if !appliedNewTip {
-		return false, false, nil, nil
+	// 5. Commit grandparent if possible. Only need to do this if we applied a new tip.
+	if appliedNewTip {
+		if err = bc.runCommitRuleOnBestChain(); err != nil {
+			return false, false, nil, errors.Wrap(err, "processBlockPoS: error committing grandparents: ")
+		}
 	}
 
-	// 5. Commit grandparent if possible.
-	if err = bc.runCommitRuleOnBestChain(); err != nil {
-		return false, false, nil, errors.Wrap(err, "processBlockPoS: error committing grandparents: ")
+	// Now that we've processed this block, we check for any orphans that
+	// are children of this block and try to process them.
+	orphansAtNextHeight := bc.blockIndexByHeight[blockNode.Height+1]
+	for _, orphan := range orphansAtNextHeight {
+		if orphan.Header.PrevBlockHash.IsEqual(blockNode.Hash) && orphan.IsStored() && !orphan.IsValidated() {
+			// TODO: how do we want to handle failures when validating orphans.
+			var orphanBlock *MsgDeSoBlock
+			orphanBlock, err = GetBlock(orphan.Hash, bc.db, bc.snapshot)
+			if err != nil {
+				glog.Errorf("processBlockPoS: Problem getting orphan block %v", orphan.Hash)
+				continue
+			}
+			if _, err = bc.validateAndIndexBlockPoS(orphanBlock); err != nil {
```

Comment:
Do we want to call processBlockPoS here instead? The goal is to extend the chain as far as possible by aggressively processing previous orphans when possible
```suggestion
			if _, err = bc.processBlockPoS(orphanBlock, currentView); err != nil {
```

Reply:
Interesting, I thought we wanted to prevent side effects like this, but I'm open to it. Need to think of any unintended consequences here. For example, what if there are multiple blocks at the next height with this block as its parent? I guess future calls to processBlockPoS will resolve the reorgs for us.

Reply:
I don't have a good answer, and would be fine wither approach here. Maybe DH has a stronger opinion here.

In either case, future reorgs triggered by the longest chain would resolve things. That said, it seems somewhat worthwhile here to do the best effort thing and see how far we can extend the chain as far as possible by aggressively and naively processing orphans. This could be especially useful if we download blocks out of order from peers, which can definitely happen.

Reply:
I think it makes sense to call `processBlockPoS` then. I've updated the PR to do so.